### PR TITLE
Clearer information about the emptyDir and correction for how topologyKey works

### PR DESCRIPTION
**Pull request by oculushut**
Minor amendment to the emptyDir documentation.  I found the existing text a little confusing.

### CODE REVIEW

**Comment by dipesh-rawat** [comment]
@oculushut Thanks you for your contribution.
Could you please sign the CLA before the PR can be reviewed.
You can follow the steps documented here: [CLA Steps](https://git.k8s.io/community/CLA.md)

File: content/en/docs/concepts/storage/volumes.md [hunk 1]
```diff
@@ -244,9 +244,8 @@ The `emptyDir.medium` field controls where `emptyDir` volumes are stored. By
 default `emptyDir` volumes are stored on whatever medium that backs the node
 such as disk, SSD, or network storage, depending on your environment. If you set
 the `emptyDir.medium` field to `"Memory"`, Kubernetes mounts a tmpfs (RAM-backed
-filesystem) for you instead.  While tmpfs is very fast, be aware that unlike
-disks, tmpfs is cleared on node reboot and any files you write count against
-your container's memory limit.
+filesystem) for you instead.  While tmpfs is very fast be aware that, unlike
+disks, files you write count against your container's memory limit.
```
Comment (sftim) [inline content/en/docs/concepts/storage/volumes.md:248]:
```suggestion
disks, files you write count against the memory limit of the container that wrote them.
```
AIUI it's even more complicated than this. If a container exists, the Pod's cgroup ends up owning the memory used to back the tmpfs. Another PR could write that up.

**Comment by tengqm** [comment]
Don't understand the merits of this change. ...

**Comment by tengqm** [comment]
> > Don't understand the merits of this change. ...
> 
> It is correcting the documentation:
> 
>     * The current example text refers to Zones which do not exist in the example configuration.  Check out the Pod Affinity page fully rendered rather than via GitHub to see the problem more clearly (https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/).
> 
>     * The text we have at the moment suggests that emptyDir disks are not cleared after a reboot - which is incorrect.  They are cleared after a reboot.  This is more obvious when you look at the full page and see that we are in the emptyDir section of the documentation.

I can get the second point, which IMHO is a clean fix. But I don't understand why the (anti-)affinity text should be removed. Are they incorrect?

**Comment by oculushut** [comment]
> > > Don't understand the merits of this change. ...
> > 
> > 
> > It is correcting the documentation:
> > ```
> > * The current example text refers to Zones which do not exist in the example configuration.  Check out the Pod Affinity page fully rendered rather than via GitHub to see the problem more clearly (https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/).
> > 
> > * The text we have at the moment suggests that emptyDir disks are not cleared after a reboot - which is incorrect.  They are cleared after a reboot.  This is more obvious when you look at the full page and see that we are in the emptyDir section of the documentation.
> > ```
> 
> I can get the second point, which IMHO is a clean fix. But I don't understand why the (anti-)affinity text should be removed. Are they incorrect?

So, go here: https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/

Search for "zone=V"

Now look at the example YAML just above where you find "zone=V".  There is no such thing as zone V in the YAML.  Similiarly, lower down there is a reference to zone R.  Again, this does not exist in the above YAML.  So, the textual description is discussing configurations that do not exist in the example and should be removed.  At least, this is my understanding unless there is some kind of implicit zone V or zone R.

**Comment by tengqm** [comment]
@oculushut The so called 'zone=V' or 'zone=R' is about pod spread topology. It implies that you are running the Pod in zone 'V', or zone 'R', and you are "supposed" to have such a label on the Node objects. With this preconditions, you can combine Pod spread topology with affinity rules.

So ... in my opinion, it is a complicated scenario, but the complexity worth it. If we revise the text, we will need to revise the YAML as well. That is, why we have a topology key in the YAML?

**Comment by oculushut** [comment]
> @oculushut The so called 'zone=V' or 'zone=R' is about pod spread topology. It implies that you are running the Pod in zone 'V', or zone 'R', and you are "supposed" to have such a label on the Node objects. With this preconditions, you can combine Pod spread topology with affinity rules.
> 
> So ... in my opinion, it is a complicated scenario, but the complexity worth it. If we revise the text, we will need to revise the YAML as well. That is, why we have a topology key in the YAML?



> @oculushut The so called 'zone=V' or 'zone=R' is about pod spread topology. It implies that you are running the Pod in zone 'V', or zone 'R', and you are "supposed" to have such a label on the Node objects. With this preconditions, you can combine Pod spread topology with affinity rules.
> 
> So ... in my opinion, it is a complicated scenario, but the complexity worth it. If we revise the text, we will need to revise the YAML as well. That is, why we have a topology key in the YAML?

I think must be missing something.  

This is the YAML that is currently published:

`apiVersion: v1
kind: Pod
metadata:
  name: with-pod-affinity
spec:
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: security
            operator: In
            values:
            - S1
        topologyKey: topology.kubernetes.io/zone
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
      - weight: 100
        podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: security
              operator: In
              values:
              - S2
          topologyKey: topology.kubernetes.io/zone
  containers:
  - name: with-pod-affinity
    image: registry.k8s.io/pause:2.0
`
How would the system know that it should put the pod in V nodes and avoid R nodes based on the above configuration?  V and R are mentioned in the descriptive text, but not that YAML example above.  So either there is a misconfiguration in the YAML or the text is wrong.  I've amended the text in my change.

Is there another configuration file on that page which shows that there are V and R nodes?

**Comment by tengqm** [comment]
> How would the system know that it should put the pod in V nodes and avoid R nodes based on the above configuration? V and R are mentioned in the descriptive text, but not that YAML example above. So either there is a misconfiguration in the YAML or the text is wrong. I've amended the text in my change.
> 
> Is there another configuration file on that page which shows that there are V and R nodes?

The `topologyKey: topology.kubernetes.io/zone` is used for this purpose. The 'V' and 'R' are not real values, they are there as examples. They are zone labels to be found on Node objects. See https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#node-labels. There are some stackoverflow discussions on this as well. https://stackoverflow.com/questions/72240224/what-is-topologykey-in-pod-affinity

We don't have a Node spec on this page, because we assumed users understand this concept. Instead of removing those texts, I'd suggest we make them more readable, more understandable.
If your concern is that the example is not a good candidate for copy-paste and experiment, we can refrain from making it an example, just a snippet in the markdown would be okay.

**Comment by Arnett28** [comment]
Ok

**Comment by oculushut** [comment]
> > How would the system know that it should put the pod in V nodes and avoid R nodes based on the above configuration? V and R are mentioned in the descriptive text, but not that YAML example above. So either there is a misconfiguration in the YAML or the text is wrong. I've amended the text in my change.
> > Is there another configuration file on that page which shows that there are V and R nodes?
> 
> The `topologyKey: topology.kubernetes.io/zone` is used for this purpose. The 'V' and 'R' are not real values, they are there as examples. They are zone labels to be found on Node objects. See https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/#node-labels. There are some stackoverflow discussions on this as well. https://stackoverflow.com/questions/72240224/what-is-topologykey-in-pod-affinity
> 
> We don't have a Node spec on this page, because we assumed users understand this concept. Instead of removing those texts, I'd suggest we make them more readable, more understandable. If your concern is that the example is not a good candidate for copy-paste and experiment, we can refrain from making it an example, just a snippet in the markdown would be okay.

I see your point now.  Perhaps something like this would make it clearer that zone V would be specified via node labels?  Esp. for anybody reading this section ahead of the zone labels section.

_The affinity rule specifies that the scheduler is allowed to place the example Pod on a node only if that node belongs to a specific [zone](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) where other Pods have been labeled with `security=S1`. For instance, if we have a cluster with a designated zone, let's call it "Zone V," consisting of nodes labeled with `topology.kubernetes.io/zone=V`, the scheduler can assign the Pod to any node within Zone V, as long as there is at least one Pod within Zone V already labeled with `security=S1`. Conversely, if there are no Pods with `security=S1` labels in Zone V, the scheduler will not assign the example Pod to any node in that zone._

**Comment by tengqm** [comment]
@oculushut The revised text looks good to me.

File: content/en/docs/concepts/scheduling-eviction/assign-pod-node.md [hunk 2]
```diff
@@ -294,13 +294,9 @@ Pod affinity rule uses the "hard"
 `requiredDuringSchedulingIgnoredDuringExecution`, while the anti-affinity rule
 uses the "soft" `preferredDuringSchedulingIgnoredDuringExecution`.
 
-The affinity rule says that the scheduler can only schedule a Pod onto a node if
-the node is in the same zone (`topologyKey: topology.kubernetes.io/zone`) as one or more existing Pods with the label
-`security=S1`.
+The affinity rule specifies that the scheduler is allowed to place the example Pod on a node only if that node belongs to a specific [zone](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) where other Pods have been labeled with `security=S1`. For instance, if we have a cluster with a designated zone, let's call it "Zone V," consisting of nodes labeled with `topology.kubernetes.io/zone=V`, the scheduler can assign the Pod to any node within Zone V, as long as there is at least one Pod within Zone V already labeled with `security=S1`. Conversely, if there are no Pods with `security=S1` labels in Zone V, the scheduler will not assign the example Pod to any node in that zone.
```
Comment (tengqm) [inline content/en/docs/concepts/scheduling-eviction/assign-pod-node.md:297]: Please manually wrap the long line for ease of change tracking and localization.
Please use relative links `[zone](/docs/...)` without hostname.

**Review by windsonsea — COMMENTED** [review]
Thanks.
/lgtm

**Pull request merged**